docs/reference/modules/discovery.asciidoc

```diff
@@ -1,30 +1,805 @@
+known master-eligible nodes by connecting to each address and attempting to
+identify the node to which it is connected. Secondly it shares with the remote
+node a list of all of its known master-eligible peers and the remote node
+responds with _its_ peers in turn. The node then probes all the new nodes that
+it just discovered, requests their peers, and so on.
+
+If the node is not master-eligible then it continues this discovery process
+until it has discovered an elected master node. If no elected master is
+discovered then the node will retry after `discovery.find_peers_interval` which
+defaults to `1s`.
+
+If the node is master-eligible then it continues this discovery process until it
+has either discovered an elected master node or else it has discovered enough
+masterless master-eligible nodes to complete an election. If neither of these
+occur quickly enough then the node will retry after
+`discovery.find_peers_interval` which defaults to `1s`.
+
+The discovery process is controlled by the following settings.
+
+`discovery.find_peers_interval`::
+
+    Sets how long a node will wait before attempting another discovery round.
+    Defaults to `1s`.
+
+`discovery.request_peers_timeout`::
+
+    Sets how long a node will wait after asking its peers again before
+    considering the request to have failed. Defaults to `3s`.
+
+`discovery.probe.connect_timeout`::
+
+    Sets how long to wait when attempting to connect to each address. Defaults
+    to `3s`.
+
+`discovery.probe.handshake_timeout`::
+
+    Sets how long to wait when attempting to identify the remote node via a
+    handshake. Defaults to `1s`.
+
+`discovery.cluster_formation_warning_timeout`::
+
+    Sets how long a node will try to form a cluster before logging a warning
+    that the cluster did not form. Defaults to `10s`.
+
+If a cluster has not formed after `discovery.cluster_formation_warning_timeout`
+has elapsed then the node will log a warning message that starts with the phrase
+`master not discovered` which describes the current state of the discovery
+process.
+
+[float]
+[[modules-discovery-bootstrap-cluster]]
+=== Bootstrapping a cluster
+
+Starting an Elasticsearch cluster for the very first time requires the initial
+set of master-eligible nodes to be explicitly set on one or more of the
+master-eligible nodes in the cluster using this setting:
+
+`cluster.initial_master_nodes`::
+
+    Sets a list of the <<node.name,node names>> or transport addresses of the
+    initial set of master-eligible nodes in a brand-new cluster. By default
+    this list is empty, meaning that this node expects to join a cluster that
+    has already been bootstrapped.
+
+This setting can be given on the command line when starting up each
+master-eligible node, or added to the `elasticsearch.yml` configuration file on
+those nodes. Once the cluster has formed this setting is no longer required and
+may be removed. It should be omitted on master-ineligible nodes, and on
```

Comment:
I still feel this suggests too strongly that it should be removed. How about:

```
Once the cluster has formed this setting is no longer required and is ignored.
```

Reply:
Ok; I added a qualification about master nodes needing persistent storage, but see 2df3878b7746bd48888219560dd3b0e950a8b477 for details.